File: frontend/src/api/tollfree.ts
import { apiClient } from './client';

export interface TollFreeMemory {
  category: string | null;
  problem: string | null;
  location: string | null;
  duration: string | null;
  affected_scope: string | null;
  frequency: string | null;
  severity: string | null;
  priority: string | null;
  department: string | null;
  citizen_name: string | null;
}

export interface TollFreeMessage {
  id: number;
  role: 'citizen' | 'ai' | 'system';
  content: string;
  normalized_content?: string | null;
  language?: string | null;
  audio_url?: string | null;
  created_at: string;
}

export interface TollFreeSessionResponse {
  session_id: string;
  caller_phone: string;
  toll_free_number: string;
  state: string;
  language: string;
  language_confidence?: number;
  current_field_prompted?: string | null;
  structured_memory: TollFreeMemory;
  complaint_id?: number | null;
  complaint_number?: string | null;
  created_at: string;
  updated_at: string;
  messages: TollFreeMessage[];
}

export interface TollFreeTurnResponse {
  success: boolean;
  session_id: string;
  state: string;
  detected_language: string;
  language_confidence?: number;
  ai_reply: string;
  spoken_reply: string;
  raw_transcript?: string;
  normalized_transcript?: string;
  memory: TollFreeMemory;
  next_field?: string;
  is_confirmation?: boolean;
  complaint_created?: boolean;
  complaint_id?: number;
  complaint_number?: string;
  department?: string;
  sms_sent?: boolean;
  unclear?: boolean;
  error_code?: string;
  message?: string;
}

export const tollfreeApi = {
  createSession: async (callerPhone = '+919843098765', tollFreeNumber = '1800-425-8693') => {
    const res = await apiClient.post('/tollfree/session', {
      caller_phone: callerPhone,
      toll_free_number: tollFreeNumber,
    });
    return res.data;
  },

  getSession: async (sessionId: string): Promise<TollFreeSessionResponse> => {
    const res = await apiClient.get(`/tollfree/session/${sessionId}`);
    return res.data;
  },

  sendAudio: async (sessionId: string, audioBlob: Blob): Promise<TollFreeTurnResponse> => {
    const formData = new FormData();
    const extension = audioBlob.type.includes('ogg') ? 'ogg' : audioBlob.type.includes('wav') ? 'wav' : 'webm';
    formData.append('audio', audioBlob, `tollfree_recording_${Date.now()}.${extension}`);

    const res = await apiClient.post(`/tollfree/session/${sessionId}/audio`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
    return res.data;
  },

  sendMessage: async (sessionId: string, message: string): Promise<TollFreeTurnResponse> => {
    const res = await apiClient.post(`/tollfree/session/${sessionId}/message`, {
      message,
    });
    return res.data;
  },

  confirmComplaint: async (sessionId: string) => {
    const res = await apiClient.post(`/tollfree/session/${sessionId}/confirm`);
    return res.data;
  },

  cancelSession: async (sessionId: string) => {
    const res = await apiClient.post(`/tollfree/session/${sessionId}/cancel`);
    return res.data;
  },

  endSession: async (sessionId: string) => {
    const res = await apiClient.post(`/tollfree/session/${sessionId}/end`);
    return res.data;
  },
};

File: backend/app/models/tollfree.py
from datetime import datetime, timezone
from typing import Optional, Dict, Any, List
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, JSON, Float
from sqlalchemy.orm import relationship
import enum

from app.database.base import Base


class TollFreeState(str, enum.Enum):
    CONNECTED = "CONNECTED"
    GREETING = "GREETING"
    WAITING_FOR_CITIZEN = "WAITING_FOR_CITIZEN"
    RECORDING = "RECORDING"
    TRANSCRIBING = "TRANSCRIBING"
    UNDERSTANDING = "UNDERSTANDING"
    AI_SPEAKING = "AI_SPEAKING"
    CONFIRMATION = "CONFIRMATION"
    REGISTERING_COMPLAINT = "REGISTERING_COMPLAINT"
    COMPLETED = "COMPLETED"
    ERROR = "ERROR"


class TollFreeCallSession(Base):
    __tablename__ = "tollfree_sessions"

    id = Column(Integer, primary_key=True, index=True)
    call_session_id = Column(String(64), unique=True, index=True, nullable=False)
    provider_call_id = Column(String(128), nullable=True, index=True)
    caller_phone = Column(String(32), nullable=True, default="+919843098765")
    toll_free_number = Column(String(32), nullable=False, default="1800-425-8693")
    
    language = Column(String(32), default="Auto", nullable=False)  # Tamil, English, Tanglish, Auto
    language_confidence = Column(Float, default=0.0)
    
    state = Column(String(64), default=TollFreeState.CONNECTED.value, nullable=False)
    current_field_prompted = Column(String(64), nullable=True)

    # Structured Conversation Memory Slots
    category = Column(String(64), nullable=True)
    problem = Column(Text, nullable=True)
    location = Column(String(255), nullable=True)
    duration = Column(String(128), nullable=True)
    affected_scope = Column(String(128), nullable=True)
    frequency = Column(String(128), nullable=True)
    severity = Column(String(64), nullable=True)
    priority = Column(String(32), nullable=True, default="MEDIUM")
    department = Column(String(128), nullable=True)
    citizen_name = Column(String(128), nullable=True)

    # Complete snapshot of memory dictionary
    structured_memory = Column(JSON, default=dict, nullable=False)

    # Full Transcript Record
    raw_transcript = Column(Text, nullable=True, default="")
    normalized_transcript = Column(Text, nullable=True, default="")

    # Real Database Complaint Linkage
    complaint_id = Column(Integer, ForeignKey("complaints.id", ondelete="SET NULL"), nullable=True)
    complaint_number = Column(String(64), nullable=True)

    # Error Tracking
    error_code = Column(String(64), nullable=True)
    error_message = Column(Text, nullable=True)

    # Timestamps
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc), nullable=False)
    ended_at = Column(DateTime, nullable=True)

    # Relationships
    messages = relationship("TollFreeMessage", back_populates="session", cascade="all, delete-orphan", order_by="TollFreeMessage.created_at")
    complaint = relationship("Complaint", foreign_keys=[complaint_id])


class TollFreeMessage(Base):
    __tablename__ = "tollfree_messages"

    id = Column(Integer, primary_key=True, index=True)
    session_id = Column(Integer, ForeignKey("tollfree_sessions.id", ondelete="CASCADE"), nullable=False, index=True)
    role = Column(String(16), nullable=False)  # "citizen", "ai", "system"
    content = Column(Text, nullable=False)
    normalized_content = Column(Text, nullable=True)
    language = Column(String(32), nullable=True)
    audio_url = Column(String(255), nullable=True)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)

    session = relationship("TollFreeCallSession", back_populates="messages")
